analyzers/src/SonarAnalyzer.CSharp/Rules/RedundancyInConstructorDestructorDeclaration.cs

```diff
@@ -18,103 +18,101 @@
-    public sealed class RedundancyInConstructorDestructorDeclaration : SonarDiagnosticAnalyzer
-    {
-        internal const string DiagnosticId = "S3253";
-        private const string MessageFormat = "Remove this redundant {0}.";
+    internal const string DiagnosticId = "S3253";
+    private const string MessageFormat = "Remove this redundant {0}.";
 
-        private static readonly SyntaxKind[] TypesWithPrimaryConstructorDeclarations =
-        {
-            SyntaxKind.ClassDeclaration,
-            SyntaxKind.StructDeclaration,
-            SyntaxKindEx.RecordClassDeclaration,
-            SyntaxKindEx.RecordStructDeclaration
-        };
+    private static readonly SyntaxKind[] TypesWithPrimaryConstructorDeclarations =
+    [
+        SyntaxKind.ClassDeclaration,
+        SyntaxKind.StructDeclaration,
+        SyntaxKindEx.RecordClassDeclaration,
+        SyntaxKindEx.RecordStructDeclaration
+    ];
 
-        private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
-        protected override void Initialize(SonarAnalysisContext context)
+    protected override void Initialize(SonarAnalysisContext context)
+    {
+        context.RegisterNodeAction(CheckConstructorDeclaration, SyntaxKind.ConstructorDeclaration);
+        context.RegisterNodeAction(CheckDestructorDeclaration, SyntaxKind.DestructorDeclaration);
+        context.RegisterNodeAction(CheckTypesWithPrimaryConstructor, TypesWithPrimaryConstructorDeclarations);
+    }
+
+    private static void CheckTypesWithPrimaryConstructor(SonarSyntaxNodeReportingContext context)
+    {
+        var typeDeclaration = (TypeDeclarationSyntax)context.Node;
+        if (!IsInheritingFromTypeWithValueInPrimaryConstructor(typeDeclaration)
+            && typeDeclaration.ParameterList() is { Parameters.Count: 0 } parameterList
+            && !IsStructWithInitializedFieldOrProperty(typeDeclaration, context.SemanticModel))
         {
-            context.RegisterNodeAction(CheckConstructorDeclaration, SyntaxKind.ConstructorDeclaration);
-            context.RegisterNodeAction(CheckDestructorDeclaration, SyntaxKind.DestructorDeclaration);
-            context.RegisterNodeAction(CheckTypesWithPrimaryConstructor, TypesWithPrimaryConstructorDeclarations);
+            context.ReportIssue(Rule, parameterList, "primary constructor");
         }
 
-        private static void CheckTypesWithPrimaryConstructor(SonarSyntaxNodeReportingContext context)
+        static bool IsInheritingFromTypeWithValueInPrimaryConstructor(TypeDeclarationSyntax node) =>
+            node.BaseList is { } baseList
```

Comment:
I did not check is type is struct or record struct because the compilation fails if one tries to inherit from another type anyways.